gds_write/def2stream.py

```diff
@@ -0,0 +1,118 @@
+parser.add_argument("-d", "--input-def",      required=True,  help="Path to def file of the implemented design after detailed routing step")
+parser.add_argument("-l", "--input-lef",      required=True,  help="Path to lef file, specify one argument for each input LEF file", action="append")
+parser.add_argument("-t", "--tech-file",      required=True,  help="Path to KLayout technology file, a mapping from LEF/DEF to GDS. Specify one argument for each input GDS file")
+parser.add_argument("-o", "--out",            required=True,  help="Path to output GDS file")
+parser.add_argument("-g", "--input-gds",      required=False, help="Paths to additional input GDS files", action="append")
+parser.add_argument("-m", "--layer-map",      required=False, help="Path to layer map file", default="")
+parser.add_argument("-e", "--gds-allow-empty",required=False, help="regex to allow empty GDS for matched cells")
+
+args = parser.parse_args()
+
+errors = 0
+
+# Load technology file
+tech = db.Technology()
+tech.load(args.tech_file)
+layoutOptions = tech.load_layout_options
+# Don't read all LEF files stored in the same directory as input DEF file
+layoutOptions.lefdef_config.read_lef_with_def = False
+# Read only those from `layoutOptions.lefdef_config.lef_files`
+# and overwrite ones specified under <lef-files></lef-files> in KLayout LYT file
+layoutOptions.lefdef_config.lef_files = args.input_lef
+# Ensure correct paths
+layoutOptions.lefdef_config.paths_relative_to_cwd = True
+if len(args.layer_map) > 0:
+  layoutOptions.lefdef_config.map_file = args.layer_map
+
+# Load def file
+main_layout = db.Layout()
+print("[INFO] Reporting cells prior to loading DEF ...")
+for i in main_layout.each_cell():
+  print("[INFO] '{0}'".format(i.name))
+
+print("[INFO] Reading DEF ...")
+main_layout.read(args.input_def, layoutOptions)
+
+# Load in the gds to merge
+print("[INFO] Merging GDS/OAS files...")
+for fil in args.input_gds:
+  print("\t{0}".format(fil))
+  main_layout.read(fil)
+
+# Copy the top level only to a new layout
+print("[INFO] Copying toplevel cell '{0}'".format(args.design_name))
+top_only_layout = db.Layout()
+top_only_layout.dbu = main_layout.dbu
+top = top_only_layout.create_cell(args.design_name)
+top.copy_tree(main_layout.cell(args.design_name))
+
+print("[INFO] Checking for missing cell from GDS/OAS...")
+missing_cell = False
+regex = None
+if args.gds_allow_empty:
+    print("[INFO] Found gds_allow_empty")
+    regex = args.gds_allow_empty
+for i in top_only_layout.each_cell():
+  if i.is_empty():
+    missing_cell = True
+    if regex is not None and re.match(regex, i.name):
+        print("[WARNING] LEF Cell '{0}' ignored. Matches gds_allow_empty.".format(i.name))
+    else:
+        print("[ERROR] LEF Cell '{0}' has no matching GDS/OAS cell."
+              " Cell will be empty.".format(i.name))
+        errors += 1
+
+if not missing_cell:
+  print("[INFO] All LEF cells have matching GDS/OAS cells")
+
+if (errors > 0):
```

Comment:
any reason we don't raise the exception in place?

Reply:
We can do that but I wanted to let the script print out all the missing cells first before raising the exception.
Let me know if you would prefer to have this script fail early.